### PR TITLE
[RemoveDIs][DebugInfo][IR] Add parsing for non-intrinsic debug values

### PR DESCRIPTION
**Pull request by SLTozer**
This patch adds support for parsing the proposed non-instruction debug info ("RemoveDIs") from textual IR. This patch does not cover bitcode parsing or documentation, and printing and verifier support are added in prior PRs (https://github.com/llvm/llvm-project/pull/79281, https://github.com/llvm/llvm-project/pull/79810); this patch is presented as a standalone patch for the sake of review, but it is intended to land after the aforementioned patches and will not compile without being based on them[0]. This patch also contains a basic affirmative test ("can we parse the new format correctly"), but not the negative tests (testing for parser errors) - these will be added soon, but shouldn't hold up review of the implementation details of the patch.

An important detail of this patch is the fact that although we can now parse in the RemoveDIs (new) and Intrinsic (old) debug info formats, we will always convert back to the old format at the end of parsing - this is done for two reasons: firstly to ensure that every tool is able to process IR printed in the new format, regardless of whether that tool has had RemoveDIs support added, and secondly to maintain the effect of the existing flags: for the tools where support for the new format has been added, we will run LLVM passes in the new format iff `--try-experimental-debuginfo-iterators=true`, and we will print in the new format iff `--write-experimental-debuginfo-iterators=true`; the format of the textual IR should have no effect on either of these features.

For reference, the format of the RemoveDIs debug info is as follows:

```
Old: call void @llvm.dbg.value(metadata i32 %a, metadata !12, metadata !DIExpression()), !dbg !14
New:   #dbg_value { i32 %a, !12, !DIExpression(), !14 }

Old: call void @llvm.dbg.declare(metadata ptr %b, metadata !13, metadata !DIExpression()), !dbg !16
New:   #dbg_declare { ptr %b, !13, !DIExpression(), !16 }

Old: call void @llvm.dbg.assign(metadata i32 %add, metadata !13, metadata !DIExpression(), metadata !15, metadata ptr %b, metadata !DIExpression()), !dbg !18
New:   #dbg_assign { i32 %add, !13, !DIExpression(), !15, ptr %b, !DIExpression(), !18 }
```

For the sake of review, a few notes on the design and implementation of this patch (some of these will be added to the docs in a subsequent patch):

- In order to support parsing this format, this patch adds `#` as a token to the lexer, and keywords for each debug record variant.
- It is an error for a module to have a mixed format, so we cannot contain calls or declarations for the existing debug intrinsics in the same module as a debug record.
- Although we print debug records with distinct indentation, it is not an error to parse a module with different indentation, just as with ordinary instructions.
- Because the set of debug records are much more limited than the set of instructions and there is a large overlap between their arguments, the parser handles all types of debug records in a single function with hardcoded parameter expectations, rather than parsing a generic list of parameters and relying on the verifier to catch incorrect parameter lists. This will likely continue even when we add other types of records, e.g. `llvm.dbg.label`.

[0] If you want to test the parser patch locally, there is a branch with all the prerequisite patches applied at: https://github.com/SLTozer/llvm-project/tree/ddd-ir-parse

### CODE REVIEW

**Review by OCHyams — COMMENTED** [review]
Code and test seem reasonable (with a few small nits scattered inline). Probably worth adding some negative (bad-input) tests, and another pair of eyes on this wouldn't hurt.

**Comment by adrian-prantl** [comment]
Let me know if this has been discussed elsewhere, but why is necessary to add a new `#` token to the grammar?

**Comment by SLTozer** [comment]
> Let me know if this has been discussed elsewhere, but why is necessary to add a new `#` token to the grammar?

The purpose of the leading `#` is to create a syntactic distinction from instructions. Without using a leading symbol to specify that a non-instruction line is beginning, the only discriminating feature would be whitespace - which is not clear enough by itself, and can't be used by the parser.
The reason why `#` specifically is used here is simply that it's free real estate - most other symbols that LLVM uses followed by letters are already classified as specific tokens (i.e. `!foo`, `@foo`, `$foo`, etc) that we'd ideally avoid overloading if alternatives are available.

**Comment by jmorse** [comment]
On top of what Stephen said, using a new token to make debug-info separate encourages everyone to consider debug-info as a first class element of LLVM. The new representation isn't either Metadata or Value based, it's the glue between the two.

**Comment by SLTozer** [comment]
I've had to put aside testing for this patch for a short while, but I'll soon be uploading about a dozen small parsing tests for the various parsing errors that can occur in this patch; meanwhile, are there are any more questions or requested changes for the functional component for this patch, or is it good to land once the tests are available and approved?

**Comment by SLTozer** [comment]
Updates added: besides adding some more tests for the negative cases as promised, and some slight fixes in the code itself, I've also updated this patch to use parentheses for debug records following the discussion on the printing patch.

File: llvm/lib/AsmParser/LLParser.cpp
Comment (danilaml) [inline llvm/lib/AsmParser/LLParser.cpp:79]: Why can't it use `UseNewDbgInfoFormat` value (and why isn't it used, if it's declared)?

Comment (SLTozer) [reply llvm/lib/AsmParser/LLParser.cpp:79]: Just due to an awkward combination of factors at the moment: not all LLVM features internally support the new debug info format yet, so we use `UseNewDbgInfoFormat` immediately before running optimization passes (the only part that fully supports the new format right now) and convert back afterwards. Right now `UseNewDbgInfoFormat` defaults to true, so if we use it here to conditionally use the new format then we'll break the tools that use the parser but don't support the new format. We might be able to change that so that `UseNewDbgInfoFormat` is only true when we use a tool that fully supports it, but that probably involves a separate patch.

In short, the declaration is the mistake here - I'll remove it.

File: llvm/test/DebugInfo/roundtrip-non-instruction-debug-info.ll
Comment (jmorse) [inline llvm/test/DebugInfo/roundtrip-non-instruction-debug-info.ll:1]: Implicit check-not's for "#dbg" too?

**Review by jmorse — COMMENTED** [review]
This is looking good, I've got some moderate nits inline.

I've got a slight twitching feeling about the addition of the debug-records as keywords (well, Identifiers) as it seems like quite a large commitment. However: we can always change the manner in which it's parsed, I think the format is good.

I'm not completely certain about making the mixing of debug-info formats a parsing error: wouldn't it be more natural to defer it until the Verifier? I get that that's impossible right now due to the conversions, but we could check for the existence of the three intrinsics by querying the `Module` in `finalizeDebugInfoFormat` and make that an error. That'd be less invasive than filtering the names of all call instruction targets. (Downside: no localised errors, but this doesn't strike me as being something hard to pin down).

Having an error on even the _declaration_ of intrinsics, not just the use of them, seems like quite a firm requirement. I'm not sure if that'll bite us, but we can downgrade the check if that proves to be a problem.

Some sadly menial nits: IMO the test names shouldn't contain "removedi" as that's referring to a transitional project name, in several years time no-one will recognise that name, IMO they should be named something like "dbg-record-invalid-$num" or something? IMHO it's also worth generating a positive-input that successfully parses so that we can point people at it and indicate the range of different things that can be parsed.

(It feels like this is all pretty close!)

File: llvm/include/llvm/AsmParser/LLToken.h
Comment (SLTozer) [reply llvm/include/llvm/AsmParser/LLToken.h:486]: I think actually these need to be moved to the "String valued tokens" above, since they follow the same model (a `DbgRecordType` token is just the text "dbg_foo").

File: llvm/lib/AsmParser/LLParser.cpp
Comment (SLTozer) [reply llvm/lib/AsmParser/LLParser.cpp:6472]: I think it's not entirely described in terms of actual token types - we refer to "Instructions" in other descriptions, for example, which aren't tokens but parseable objects, i.e. anything that we can call `parseFoo` for - but we can follow that description more closely than I've done here, since we do for example call `parseMetadata` for the "value" arguments.

**Comment by jmorse** [comment]
For the mixing-old-and-new issue, AFAIUI the population of people hand-writing debug-info IR tests is small, and we usually copy existing tests to write them. It's also not a syntactic problem to report as parsing will succeed, but the Module will be ill formed, it's more akin to a Verifier error but it makes the most sense to detect it in the parser (to avoid answering questions about what mixture of old/new debug-info should be permitted).

IMO: it'd be better to print out an informative error at the end of parsing, something like "#dbg_* records were encountered during parsing, but llvm.dbg.* intrinsics were seen in functions [list first ten users of the relevant declaration?]. Cannot load a module with both flavours of debug-info". That avoids the invasive code changes but gives a good diagnosis of what's gone wrong.

**Comment by SLTozer** [comment]
> It's also not a syntactic problem to report as parsing will succeed, but the Module will be ill formed, it's more akin to a Verifier error but it makes the most sense to detect it in the parser

SGTM, (this was the first way I implemented it anyway!)

File: llvm/lib/AsmParser/LLParser.cpp
Comment (danilaml) [reply llvm/lib/AsmParser/LLParser.cpp:79]: Ok!

File: llvm/test/Verifier/RemoveDI/blockbyref.ll
Comment (OCHyams) [inline llvm/test/Verifier/RemoveDI/blockbyref.ll:1]: nit: don't need `<`? and while you're there can you add a space between `1` and `|`?

EDIT: this is in a few of the other tests too. Were these copied and modified from existing tests? Feel free to ignore this nit

File: llvm/test/Verifier/RemoveDI/llvm.dbg.intrinsic-dbg-attachment.ll
Comment (OCHyams) [inline llvm/test/Verifier/RemoveDI/llvm.dbg.intrinsic-dbg-attachment.ll:9]: what's invalid about the "location" and which of the 3 locations (position, src-loc, var-loc) is this referring to?

**Review by OCHyams — COMMENTED** [review]
Latest code + test changes LGTM

@jmorse might want to check he's happy with the overall approach now though

File: llvm/test/Verifier/RemoveDI/llvm.dbg.intrinsic-dbg-attachment.ll
Comment (SLTozer) [reply llvm/test/Verifier/RemoveDI/llvm.dbg.intrinsic-dbg-attachment.ll:9]: The location overload strikes again... I'll update the verifier check to state more clearly that it's a DILocation.

**Comment by jmorse** [comment]
> @jmorse might want to check he's happy with the overall approach now though

I'm fine with it, although we should be ready to adjust it if there's turbulence.

File: llvm/test/Verifier/RemoveDI/llvm.dbg.intrinsic-dbg-attachment.ll
Comment (OCHyams) [reply llvm/test/Verifier/RemoveDI/llvm.dbg.intrinsic-dbg-attachment.ll:9]: SGTM thanks

**Review by OCHyams — APPROVED** [review]

**Pull request merged**
